### ISSUE TITLE
Design: Multi-model UI awareness signal pipeline

### ISSUE THREAD

**Issue by kevintruong** [adjienjeknwc/cogniflow#1]
SPUR design epic for the multi-model UI awareness signal pipeline.

Plan ID: 8f05d605-d84e-480b-8543-677e98167cd5

Problem
SUM UI is useful for aggregate awareness but insufficient as a standalone effectiveness judgment. The design needs distinct signals for spatial awareness, temporal visual flow, and broader human response.

Current candidate architecture
1. Static awareness: SUM UI plus DeepGaze IIE.
2. Temporal flow: EyeFormer candidate, with DeepGaze III retained as fallback and comparison baseline.
3. Human response and preference: UniAR candidate in shadow evaluation only.
4. AI reasoning: preserve model provenance and expose disagreement; do not collapse correlated outputs into one score.

Approval gate
No implementation begins until the architecture and benchmark acceptance criteria are approved.

Tracking limitation
The required dynamic label spur:plan-id:8f05d605-d84e-480b-8543-677e98167cd5 could not be created because the available GitHub token lacks label-write permission. The plan ID is retained here as the stable identifier.

**Comment by kevintruong** [comment]
[[spur-audit v1]]
event: design-research-update
plan_id: 8f05d605-d84e-480b-8543-677e98167cd5
decision: propose-shadow-first

EyeFormer finding
- Strong fit for the temporal visual-flow axis: it predicts ordered fixations and duration, supports population and personalized modes, and was evaluated on GUI data.
- Official code and checkpoints exist.
- Production gates: legacy Python 3.6, PyTorch 1.7.1, CUDA 11.0 stack; fixed-window scanpaths; repository has no visible license.
- Recommendation: benchmark as the primary temporal candidate, while keeping DeepGaze III as fallback and baseline until latency, compatibility, licensing, and calibration pass.

UniAR finding
- Strong conceptual fit for the broader response/preference axis: prompted outputs cover saliency, importance, scanpath, and aesthetic or quality ratings.
- Its heatmap and scanpath heads overlap SUM, IIE, and EyeFormer, so treating them as independent votes would double-count correlated evidence.
- Production gates: the paper describes an 848M-parameter JAX model trained on 64 TPU v3s; no official public implementation or checkpoint was found in the authors official paper or Google Research page.
- Recommendation: define the response contract now, but run UniAR only as a non-blocking research/shadow evaluator until artifacts, licensing, latency, and UI-specific calibration are resolved.

Proposed output axes
- awareness_coverage
- action_visibility
- flow_to_target
- response_or_preference
- model_disagreement

Acceptance direction
Use task-specific ground truth and per-axis thresholds. Never ship a single blended effectiveness score without provenance and calibration.

**Comment by kevintruong** [comment]
[[spur-audit v1]]
{
  "kind": "plan-submit",
  "plan_id": "8f05d605-d84e-480b-8543-677e98167cd5",
  "decision": "Signal contract separates raw model evidence from target-aware derived metrics",
  "rationale": "Heatmaps and scanpaths do not directly answer product questions. Each artifact must be evaluated against declared UI regions and intended flow, retain model provenance, and expose non-claims.",
  "axes": {
    "awareness_coverage": "SUM UI: which regions receive aggregate attention and how concentrated or distracted it is",
    "action_visibility": "DeepGaze IIE: whether intended actions are visually salient in a static first-glance cross-check",
    "flow_to_target": "EyeFormer primary and DeepGaze III baseline: expected fixation order, time or steps to target, dwell, loops, and detours",
    "response_or_preference": "UniAR importance and rating heads: whether attention aligns with perceived importance and likely aesthetic or quality response",
    "model_disagreement": "cross-model ROI ranks, target misses, and applicability warnings"
  },
  "rule": "No raw output or blended score may be interpreted as comprehension, task success, conversion, or an individual users preference without human validation."
}

**Comment by kevintruong** [comment]
[[spur-audit v1]]
{
  "kind": "plan-submit",
  "plan_id": "8f05d605-d84e-480b-8543-677e98167cd5",
  "decision": "Refine the proposed render-evaluate-repair loop into a provenance-preserving, tri-state hybrid architecture",
  "rationale": "A screenshot alone loses stable element identity and declared intent; learned signals are uncertain and cannot be fed directly to hard Z3 assertions; iterative repair needs regression checks and termination guards.",
  "architecture_changes": [
    "Replace Screenshot I_t with RenderBundle_t containing screenshot, viewport and DPR, DOM or semantics tree, element bounding boxes, interaction state, and declared goals.",
    "Treat UMSI++ and SUM UI as candidates for the same UI-specific static saliency slot and select by an in-domain benchmark rather than count both as independent votes.",
    "Run DeepGaze IIE as general static baseline; EyeFormer as UI-temporal candidate; DeepGaze III as temporal fallback and disagreement baseline.",
    "Add calibration and element-level projection before constructing the Attention Graph.",
    "Use VLM semantics as evidence over screenshot plus structure, not as the sole source of element identity or intent.",
    "Separate hard deterministic rules from uncertain soft objectives; pass only thresholded facts with provenance into solver gates.",
    "Replace binary PASS with PASS, REVISE, and REVIEW or INCONCLUSIVE.",
    "Bound repair with change budget, max iterations, cycle detection, compile and render validation, delta evaluation against I_t, best-state retention, and rollback on regression."
  ],
  "recommended_approach": "Hybrid evidence graph plus hard-rule solver and soft-objective optimizer"
}

**Comment by kevintruong** [comment]
[[spur-audit v1]]
{
  "kind": "plan-submit",
  "plan_id": "8f05d605-d84e-480b-8543-677e98167cd5",
  "decision": "EyeFormer plus UniAR is the preferred complementary research pair, not a standalone production oracle",
  "rationale": "EyeFormer contributes UI-oriented temporal order, duration, and personalization; UniAR contributes saliency, importance, scanpath, and relative rating. The pair still lacks authoritative UI intent, semantic structure, accessibility and interaction checks, independent calibration, actual task-success evidence, and production-ready UniAR artifacts.",
  "recommended_mode": {
    "eyeformer": "primary temporal-flow evidence",
    "uniar": "importance and relative-response evidence; overlapping heatmap and scanpath heads remain shadow comparisons",
    "guardrails": "declared goals plus DOM or semantics, hard specifications, disagreement state, and human benchmark calibration"
  }
}

**Comment by kevintruong** [comment]
[[spur-audit v1]]
{
  "kind": "plan-submit",
  "plan_id": "8f05d605-d84e-480b-8543-677e98167cd5",
  "decision": "User approved the experiment sequence; authoritative written spec created for review",
  "spec": "docs/superpowers/specs/2026-08-28-modal-visual-evaluation-design.ipynb",
  "commit": "3b22cd6",
  "status": "awaiting-written-spec-review",
  "formal_cells": [
    {
      "spec_id": "MODAL-EXPERIMENT-ROUTING",
      "cell_id": "f4715a2e-8cae-4c6a-9b20-000000000004",
      "report_hash": "5c308f34ff7c51f016abecfba1e24fb2d62bed515114a9f9bb598be3ccbfe781",
      "verified": true,
      "proof_fresh": true
    },
    {
      "spec_id": "DESIGN-EVALUATION-DECISION",
      "cell_id": "f4715a2e-8cae-4c6a-9b20-000000000007",
      "report_hash": "5a1a65cd40df845cd759ca4c4aa485c42cf6996f8a7365a8193358237e573a2e",
      "verified": true,
      "proof_fresh": true
    },
    {
      "spec_id": "CANDIDATE-ACCEPTANCE",
      "cell_id": "f4715a2e-8cae-4c6a-9b20-000000000009",
      "report_hash": "0301bdbe159f428b13c2f4603b3fe81431743c13ebc4e5583f60ce71a126363d",
      "verified": true,
      "proof_fresh": true
    }
  ],
  "review": {
    "placeholders": 0,
    "formal_cells": 3,
    "all_obligations_matched": true,
    "existing_modal_app_untouched": true
  },
  "authoring_note": "The skill referenced ns-spec-notebook-authoring.md was absent from the projected skill package. The live notebook_ns_mermaid_spec registry was used as the dialect authority; every cell was preflighted and executed."
}

**Comment by kevintruong** [comment]
Written specification approved by the project owner for controlled evaluation.\n\nAuthoritative spec: docs/superpowers/specs/2026-08-28-modal-visual-evaluation-design.ipynb\nSpec commit: 3b22cd6\n\nFormal evidence:\n- MODAL-EXPERIMENT-ROUTING — cell f4715a2e-8cae-4c6a-9b20-000000000004 — proof 5c308f34ff7c51f016abecfba1e24fb2d62bed515114a9f9bb598be3ccbfe781\n- DESIGN-EVALUATION-DECISION — cell f4715a2e-8cae-4c6a-9b20-000000000007 — proof 5a1a65cd40df845cd759ca4c4aa485c42cf6996f8a7365a8193358237e573a2e\n- CANDIDATE-ACCEPTANCE — cell f4715a2e-8cae-4c6a-9b20-000000000009 — proof 0301bdbe159f428b13c2f4603b3fe81431743c13ebc4e5583f60ce71a126363d\n\nScope decision: implementation planning starts with the isolated EyeFormer technical evaluation. ShareLab-SII/UniAR-RL remains a later, separately gated semantics/generation experiment. The existing cogniflow-saliency Modal app remains untouched.

**Comment by kevintruong** [comment]
Scope amendment approved by the project owner on 2026-08-28: ShareLab-SII/UniAR-RL is cancelled because its understanding/generation purpose does not match the required human-attention signal role. It has no successor task, endpoint, model download, volume, or GPU allocation in this roadmap.\n\nEyeFormer remains the only new-model evaluation. SUM UI, DeepGaze IIE, and DeepGaze III remain comparison providers.\n\nUpdated authoritative spec and implementation plan: commit f51a86e.\nUpdated MODAL-EXPERIMENT-ROUTING proof: cell f4715a2e-8cae-4c6a-9b20-000000000004, report hash 530bdb33b5bcde0e800587a6137d1ba8b7ff76b1f234c7422ac57512c2743492. The verified route now contains only run_eyeformer and hold.
